mypy/server/update.py

```diff
@@ -1,49 +1,119 @@
-"""Update build result by incrementally processing changed modules.
+"""Update build by processing changes using fine-grained dependencies.
 
 Use fine-grained dependencies to update targets in other modules that
 may be affected by externally-visible changes in the changed modules.
 
-Terms:
+This forms the core of the fine-grained incremental daemon mode. This
+module is not used at all by the 'classic' (non-daemon) incremental
+mode.
 
-* A 'target' is a function definition or the top level of a module. We
-  refer to targets using their fully qualified name (e.g. 'mod.Cls.attr').
-  Targets are the smallest units of processing during fine-grained
-  incremental checking.
-* A 'trigger' represents the properties of a part of a program, and it
-  gets triggered/activated when these properties change. For example,
-  '<mod.func>' refers to a module-level function, and it gets triggered
-  if the signature of the function changes, or if if the function is
-  removed.
+Here is some motivation for this mode:
 
-Some program state is maintained across multiple build increments:
+* By keeping program state in memory between incremental runs, we
+  only have to process changed modules, not their dependencies. The
+  classic incremental mode has to deserialize the symbol tables of
+  all dependencies of changed modules, which can be slow for large
+  programs.
 
-* The full ASTs of all modules in memory all the time (+ type map).
-* Maintain a fine-grained dependency map, which is from triggers to
-  targets/triggers. The latter determine what other parts of a program
-  need to be processed again due to an externally visible change to a
-  module.
+* Fine-grained dependencies allow processing only the relevant parts
+  of modules indirectly affected by a change. Say, if only one function
+  in a large module is affected by a change in another module, only this
+  function is processed. The classic incremental mode always processes
+  an entire file as a unit, which is typically much slower.
 
-We perform a fine-grained incremental program update like this:
+* It's possible to independently process individual modules within an
+  import cycle (SCC). Small incremental changes can be fast independent
+  of the size of the related SCC. In classic incremental mode, any change
```

Comment:
As I understand, there still can be cycles in fine grained dependencies (like mutually recursive functions). So there should be a similar notion of an SCC made of smaller locations (functions instead of modules.) Or am I missing something?

Reply:
The concept of SCC is currently implicit in fine-grained incremental mode. We can freely follow fine-grained dependencies independent of import relationships, since everything is in memory all the time. We can more or less treat the entire program as a single SCC. In case of module `x` depending on types inferred in module `y` and vice versa, we may need to perform multiple passes of type checking across the entire stale set of targets, I think. This isn't implemented yet, though.

Also, some dependency cycles are naturally handled by just following fine-grained dependencies. Say, `f` and `g` are mutually recursive, defined in different modules, and both signatures change. We first update module containing `f` and its signature changes. This causes `g` to be stale and we reprocess it (using the old definition for `g`). Then we update the module containing `g` and its signature changes as well. This triggers another update of `f`, and after than everything should be consistent.